### DISCUSSION TITLE
Schema Migration with Flyway

### DISCUSSION THREAD

**Discussion by singhbaljit** [spring-projects/spring-batch#4559] [Q&A] [answered]
I understand that `spring.batch.jdbc.initialize-schema` provides the ability for Spring Batch to manage DB schema. However, my project already uses Flyway. Spring Boot recommends disabling other migrations when using Flyway (or Liquidbase). I do see that the `spring.batch.jdbc.schema` refers to the migration scripts in [core](https://github.com/spring-projects/spring-batch/tree/main/spring-batch-core/src/main/resources/org/springframework/batch/core). However, these files are very old (2+ years). I also see some newer files in the `migration` directory. I've a few questions/requests:

1. What is the best way to migrate schema when using a higher-level migration tool like Flyway? It appears the either documentation or implementation is conflicting.
2. If the answer is that Flyway should be manage, then what is the single source of truth for the schema that I can use in my Flyway migration script? Can this be maintained and released alongside releases so people can migrate DBs with newer versions of Batch?

**Comment by prafsoni** [comment]
It would be great if support for liquibase is also added along side flyway

**Comment by fmbenhassine** [comment, marked as answer]
The `core` package in every versioned `spring-batch-core` jar contains the DDL scripts that can be applied without modification for new apps using that version (ie creating the tables from scratch). For existing apps migrating from a prior version, one should apply the DDL scripts in the `migration` package up until the current version.

For example, if you are deploying a new batch app using v5.0, you can apply the scripts from the `core` package of the v5.0 jar. If you are migrating an existing app from v4.3 to v5.0 (ie the tables were already created from the v4.3 jar), then you need to apply the scripts from the `migration/5.0` package of the v5.0 jar on top of existing tables. There is a note about this in the docs [here](https://docs.spring.io/spring-batch/reference/schema-appendix.html#migrationDDLScripts).

Now the way to apply these scripts (manually or with a migration tool) is out of the scope of Spring Batch. Hopefully this answers your questions.

**Comment by fmbenhassine** [comment]
I believe my last comment answers the question, so I am closing this discussion. But if you need more help, please add a comment.
